File: onedal/linear_model/linear_model.py
# ==============================================================================
# Copyright 2023 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

from abc import ABCMeta, abstractmethod
from numbers import Number

import numpy as np

from daal4py.sklearn._utils import daal_check_version, get_dtype, make2d

from .._config import _get_config
from ..common._base import BaseEstimator
from ..common._estimator_checks import _check_is_fitted
from ..common.hyperparameters import get_hyperparameters
from ..datatypes import from_table, to_table
from ..utils import _check_array, _check_n_features, _check_X_y, _num_features
from ..utils._array_api import _get_sycl_namespace


class BaseLinearRegression(BaseEstimator, metaclass=ABCMeta):
    """
    Base class for LinearRegression oneDAL implementation.
    """

    @abstractmethod
    def __init__(self, fit_intercept, copy_X, algorithm, alpha=0.0):
        self.fit_intercept = fit_intercept
        self.alpha = alpha
        self.copy_X = copy_X
        self.algorithm = algorithm

    def _get_onedal_params(self, dtype=np.float32):
        intercept = "intercept|" if self.fit_intercept else ""
        params = {
            "fptype": dtype,
            "method": self.algorithm,
            "intercept": self.fit_intercept,
            "result_option": (intercept + "coefficients"),
        }
        if daal_check_version((2024, "P", 600)):
            params["alpha"] = self.alpha

        return params

    def _create_model(self, policy):
        module = self._get_backend("linear_model", "regression")
        model = module.model()

        coefficients = self.coef_
        dtype = get_dtype(coefficients)

        if coefficients.ndim == 2:
            n_features_in = coefficients.shape[1]
            n_targets_in = coefficients.shape[0]
        else:
            n_features_in = coefficients.size
            n_targets_in = 1

        intercept = self.intercept_
        if isinstance(intercept, Number):
            assert n_targets_in == 1
        else:
            if not isinstance(intercept, np.ndarray):
                intercept = np.asarray(intercept, dtype=dtype)
            assert n_targets_in == intercept.size

        coefficients, intercept = make2d(coefficients), make2d(intercept)
        coefficients = coefficients.T if n_targets_in == 1 else coefficients

        assert coefficients.shape == (
            n_targets_in,
            n_features_in,
        ), f"{coefficients.shape}, {n_targets_in}, {n_features_in}"
        assert intercept.shape == (n_targets_in, 1), f"{intercept.shape}, {n_targets_in}"

        desired_shape = (n_targets_in, n_features_in + 1)
        packed_coefficients = np.zeros(desired_shape, dtype=dtype)

        packed_coefficients[:, 1:] = coefficients
        if self.fit_intercept:
            packed_coefficients[:, 0][:, np.newaxis] = intercept

        model.packed_coefficients = to_table(
            packed_coefficients, queue=getattr(policy, "_queue", None)
        )

        self._onedal_model = model

        return model

    def predict(self, X, queue=None):
        """
        Predict using the linear model.
        Parameters
        ----------
        X : array-like or sparse matrix, shape (n_samples, n_features)
            Samples.

        queue : dpctl.SyclQueue
            If not None, uses this queue for computations.

        Returns
        -------
        C : array, shape (n_samples, n_targets)
            Returns predicted values.
        """
        module = self._get_backend("linear_model", "regression")

        _check_is_fitted(self)

        sua_iface, xp, _ = _get_sycl_namespace(X)
        use_raw_input = _get_config().get("use_raw_input") is True

        policy = self._get_policy(queue, X)

        if not use_raw_input:
            X = _check_array(
                X, dtype=[np.float64, np.float32], force_all_finite=False, ensure_2d=False
            )
            X = make2d(X)

        _check_n_features(self, X, False)

        if hasattr(self, "_onedal_model"):
            model = self._onedal_model
        else:
            model = self._create_model(policy)

        X_table = to_table(X, queue=queue)
        params = self._get_onedal_params(X_table.dtype)

        result = module.infer(policy, params, model, X_table)
        y = from_table(result.responses, sua_iface=sua_iface, sycl_queue=queue, xp=xp)

        if y.shape[1] == 1 and self.coef_.ndim == 1:
            return xp.reshape(y, (-1,))
        else:
            return y


class LinearRegression(BaseLinearRegression):
    """
    Linear Regression oneDAL implementation.

    Parameters
    ----------
    fit_intercept : bool, default=True
        Whether to calculate the intercept for this model. If set
        to False, no intercept will be used in calculations
        (i.e. data is expected to be centered).

    copy_X : bool, default=True
        If True, X will be copied; else, it may be overwritten.

    algorithm : string, default="norm_eq"
        Algorithm used for computation on oneDAL side
    """

    def __init__(
        self,
        fit_intercept=True,
        copy_X=False,
        *,
        algorithm="norm_eq",
        **kwargs,
    ):
        super().__init__(fit_intercept=fit_intercept, copy_X=copy_X, algorithm=algorithm)

    def fit(self, X, y, queue=None):
        """
        Fit linear model.
        Parameters
        ----------
        X : {array-like, sparse matrix} of shape (n_samples, n_features)
            Training data.

        y : array-like of shape (n_samples,) or (n_samples, n_targets)
            Target values. Will be cast to X's dtype if necessary.

        queue : dpctl.SyclQueue
            If not None, use this queue for computations.

        Returns
        -------
        self : object
            Fitted Estimator.
        """
        module = self._get_backend("linear_model", "regression")

        sua_iface, xp, _ = _get_sycl_namespace(X)
        use_raw_input = _get_config().get("use_raw_input") is True
        if use_raw_input and sua_iface is not None:
            queue = X.sycl_queue

        if not use_raw_input:
            # TODO Fix _check_X_y to make sure this conversion is there
            if not isinstance(X, np.ndarray):
                X = np.asarray(X)

            dtype = get_dtype(X)
            if dtype not in [np.float32, np.float64]:
                dtype = np.float64
                X = X.astype(dtype, copy=self.copy_X)

            y = np.asarray(y).astype(dtype=dtype)

            X, y = _check_X_y(X, y, force_all_finite=False, accept_2d_y=True)

        policy = self._get_policy(queue, X, y)

        self.n_features_in_ = _num_features(X, fallback_1d=True)

        X_table, y_table = to_table(X, y, queue=queue)
        params = self._get_onedal_params(X_table.dtype)

        hparams = get_hyperparameters("linear_regression", "train")
        if hparams is not None and not hparams.is_default:
            result = module.train(policy, params, hparams.backend, X_table, y_table)
        else:
            result = module.train(policy, params, X_table, y_table)

        self._onedal_model = result.model

        packed_coefficients = from_table(
            result.model.packed_coefficients, sua_iface=sua_iface, sycl_queue=queue, xp=xp
        )
        self.coef_, self.intercept_ = (
            packed_coefficients[:, 1:],
            packed_coefficients[:, 0],
        )

        if self.coef_.shape[0] == 1 and y.ndim == 1:
            self.coef_ = xp.reshape(self.coef_, (-1,))
            self.intercept_ = self.intercept_[0]

        return self


class Ridge(BaseLinearRegression):
    """
    Ridge Regression oneDAL implementation.

    Parameters
    ----------
    alpha : float, default=1.0
        Regularization strength; must be a positive float. Regularization
        improves the conditioning of the problem and reduces the variance of
        the estimates. Larger values specify stronger regularization.

    fit_intercept : bool, default=True
        Whether to calculate the intercept for this model. If set
        to False, no intercept will be used in calculations
        (i.e. data is expected to be centered).

    copy_X : bool, default=True
        If True, X will be copied; else, it may be overwritten.

    algorithm : string, default="norm_eq"
        Algorithm used for computation on oneDAL side.
    """

    def __init__(
        self,
        alpha=1.0,
        fit_intercept=True,
        copy_X=False,
        *,
        algorithm="norm_eq",
        **kwargs,
    ):
        super().__init__(
            fit_intercept=fit_intercept, alpha=alpha, copy_X=copy_X, algorithm=algorithm
        )

    def fit(self, X, y, queue=None):
        """
        Fit linear model.
        Parameters
        ----------
        X : {array-like, sparse matrix} of shape (n_samples, n_features)
            Training data.

        y : array-like of shape (n_samples,) or (n_samples, n_targets)
            Target values. Will be cast to X's dtype if necessary.

        queue : dpctl.SyclQueue
            If not None, use this queue for computations.

        Returns
        -------
        self : object
            Fitted Estimator.
        """
        sua_iface, xp, _ = _get_sycl_namespace(X)

        module = self._get_backend("linear_model", "regression")
        _, xp, _ = _get_sycl_namespace(X)

        if not isinstance(X, np.ndarray):
            X = np.asarray(X)

        dtype = get_dtype(X)
        if dtype not in [np.float32, np.float64]:
            dtype = np.float64
            X = X.astype(dtype, copy=self.copy_X)

        use_raw_input = _get_config().get("use_raw_input") is True
        if not use_raw_input:
            X = _check_array(
                X,
                dtype=[np.float64, np.float32],
                force_all_finite=False,
                ensure_2d=False,
                copy=self.copy_X,
            )
            X, y = _check_X_y(X, y, force_all_finite=False, accept_2d_y=True)
            y = np.asarray(y).astype(dtype=get_dtype(X))

        policy = self._get_policy(queue, X, y)

        self.n_features_in_ = _num_features(X, fallback_1d=True)

        X_table, y_table = to_table(X, y, queue=queue)
        params = self._get_onedal_params(X.dtype)

        result = module.train(policy, params, X_table, y_table)
        self._onedal_model = result.model

        packed_coefficients = from_table(
            result.model.packed_coefficients, sua_iface=sua_iface, sycl_queue=queue, xp=xp
        )
        self.coef_, self.intercept_ = (
            packed_coefficients[:, 1:],
            packed_coefficients[:, 0],
        )

        if self.coef_.shape[0] == 1 and y.ndim == 1:
            self.coef_ = xp.reshape(self.coef_, (-1,))
            self.intercept_ = self.intercept_[0]

        return self

File: onedal/linear_model/linear_model.cpp
/*******************************************************************************
* Copyright 2023 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

#include "oneapi/dal/algo/linear_regression.hpp"

#include "onedal/common.hpp"
#include "onedal/version.hpp"

#include <regex>

namespace py = pybind11;

namespace oneapi::dal::python {

namespace linear_model {

template <typename Task, typename Ops>
struct method2t {
    method2t(const Task& task, const Ops& ops) : ops(ops) {}

    template <typename Float>
    auto operator()(const py::dict& params) {
        using namespace dal::linear_regression;

        const auto method = params["method"].cast<std::string>();
        ONEDAL_PARAM_DISPATCH_VALUE(method, "norm_eq", ops, Float, method::norm_eq);
        ONEDAL_PARAM_DISPATCH_VALUE(method, "by_default", ops, Float, method::norm_eq);
        ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(method);
    }

    Ops ops;
};

auto get_onedal_result_options(const py::dict& params) {
    using namespace dal::linear_regression;

    auto result_option = params["result_option"].cast<std::string>();
    result_option_id onedal_options;

    try {
        std::regex re("\\w+");
        const std::sregex_iterator last{};
        const std::sregex_iterator first( //
            result_option.begin(),
            result_option.end(),
            re);

        for (std::sregex_iterator it = first; it != last; ++it) {
            std::smatch match = *it;
            if (match.str() == "intercept") {
                onedal_options = onedal_options | result_options::intercept;
            }
            else if (match.str() == "coefficients") {
                onedal_options = onedal_options | result_options::coefficients;
            }
            else {
                ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(result_option);
            }
        }
    }
    catch (std::regex_error& e) {
        ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(result_option);
    }

    return onedal_options;
}

struct params2desc {
    template <typename Float, typename Method, typename Task>
    auto operator()(const py::dict& params) {
        using namespace dal::linear_regression;

        const auto intercept = params["intercept"].cast<bool>();

        auto desc = linear_regression::descriptor<Float, Method, Task>(intercept)
            .set_result_options(get_onedal_result_options(params));
        return desc;
    }
};

template <typename Policy, typename Task>
void init_train_ops(py::module& m) {
#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240000
    using train_hyperparams_t = dal::linear_regression::detail::train_parameters<Task>;
    m.def("train", [](
        const Policy& policy,
        const py::dict& params,
        const train_hyperparams_t& hyperparams,
        const table& data,
        const table& responses) {
            using namespace dal::linear_regression;
            using input_t = train_input<Task>;
            train_ops_with_hyperparams ops(
                policy, input_t{ data, responses }, params2desc{}, hyperparams);
            return fptype2t{ method2t{ Task{}, ops } }(params);
        }
    );
#endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240000
    m.def("train", [](
        const Policy& policy,
        const py::dict& params,
        const table& data,
        const table& responses) {
            using namespace dal::linear_regression;
            using input_t = train_input<Task>;
            train_ops ops(policy, input_t{ data, responses }, params2desc{});
            return fptype2t{ method2t{ Task{}, ops } }(params);
        }
    );
};

template <typename Policy, typename Task>
void init_partial_train_ops(py::module& m) {
    using prev_result_t = dal::linear_regression::partial_train_result<Task>;
    using train_hyperparams_t = dal::linear_regression::detail::train_parameters<Task>;
    m.def("partial_train", [](
        const Policy& policy,
        const py::dict& params,
        const train_hyperparams_t& hyperparams,
        const prev_result_t& prev,
        const table& data,
        const table& responses) {
            using namespace dal::linear_regression;
            using input_t = partial_train_input<Task>;
            partial_train_ops_with_hyperparams ops(
                policy, input_t{ prev, data, responses }, params2desc{}, hyperparams);
            return fptype2t{ method2t{ Task{}, ops } }(params);
        }
    );

    using prev_result_t = dal::linear_regression::partial_train_result<Task>;
    m.def("partial_train", [](
        const Policy& policy,
        const py::dict& params,
        const prev_result_t& prev,
        const table& data,
        const table& responses) {
            using namespace dal::linear_regression;
            using input_t = partial_train_input<Task>;
            partial_train_ops ops(policy, input_t{ prev, data, responses }, params2desc{});
            return fptype2t{ method2t{ Task{}, ops } }(params);
        }
    );
};

template <typename Policy, typename Task>
void init_finalize_train_ops(py::module& m) {
    using input_t = dal::linear_regression::partial_train_result<Task>;
    using train_hyperparams_t = dal::linear_regression::detail::train_parameters<Task>;

    m.def("finalize_train", [](
        const Policy& policy,
        const py::dict& params,
        const train_hyperparams_t& hyperparams,
        const input_t& data) {
            finalize_train_ops_with_hyperparams ops(policy, data, params2desc{}, hyperparams);
            return fptype2t{ method2t{ Task{}, ops } }(params);
        }
    );

    m.def("finalize_train", [](
        const Policy& policy,
        const py::dict& params,
        const input_t& data) {
            finalize_train_ops ops(policy, data, params2desc{});
            return fptype2t{ method2t{ Task{}, ops } }(params);
        }
    );
};

template <typename Policy, typename Task>
void init_infer_ops(py::module_& m) {
    m.def("infer",
          [](const Policy& policy,
             const py::dict& params,
             const dal::linear_regression::model<Task>& model,
             const table& data) {
              using namespace dal::linear_regression;
              using input_t = infer_input<Task>;

              infer_ops ops(policy, input_t{ data, model }, params2desc{});
              return fptype2t{ method2t{ Task{}, ops } }(params);
          });
}

template <typename Task>
void init_model(py::module_& m) {
    using namespace dal::linear_regression;
    using model_t = model<Task>;

    auto cls = py::class_<model_t>(m, "model")
                   .def(py::init())
                   .def(py::pickle(
                       [](const model_t& m) {
                           return serialize(m);
                       },
                       [](const py::bytes& bytes) {
                           return deserialize<model_t>(bytes);
                       }))
                   .DEF_ONEDAL_PY_PROPERTY(packed_coefficients, model_t);
}

template <typename Task>
void init_train_result(py::module_& m) {
    using namespace dal::linear_regression;
    using result_t = train_result<Task>;

    py::class_<result_t>(m, "train_result")
        .def(py::init())
        .DEF_ONEDAL_PY_PROPERTY(model, result_t)
        .DEF_ONEDAL_PY_PROPERTY(intercept, result_t)
        .DEF_ONEDAL_PY_PROPERTY(coefficients, result_t)
        .DEF_ONEDAL_PY_PROPERTY(packed_coefficients, result_t)
        .DEF_ONEDAL_PY_PROPERTY(result_options, result_t);
}

template <typename Task>
void init_partial_train_result(py::module_& m) {
    using namespace dal::linear_regression;
    using result_t = partial_train_result<Task>;
    py::class_<result_t>(m, "partial_train_result")
        .def(py::init())
        .DEF_ONEDAL_PY_PROPERTY(partial_xtx, result_t)
        .DEF_ONEDAL_PY_PROPERTY(partial_xty, result_t);
}

template <typename Task>
void init_infer_result(py::module_& m) {
    using namespace dal::linear_regression;
    using result_t = infer_result<Task>;

    auto cls = py::class_<result_t>(m, "infer_result")
                   .def(py::init())
                   .DEF_ONEDAL_PY_PROPERTY(responses, result_t);
}

#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240000

template <typename Task>
void init_train_hyperparameters(py::module_& m) {
    using namespace dal::linear_regression::detail;
    using train_hyperparams_t = train_parameters<Task>;

    auto cls = py::class_<train_hyperparams_t>(m, "train_hyperparameters")
                   .def(py::init())
                   .def("set_cpu_macro_block", [](train_hyperparams_t& self, int64_t cpu_macro_block) {
                        self.set_cpu_macro_block(cpu_macro_block);
                   })
                   .def("set_gpu_macro_block", [](train_hyperparams_t& self, int64_t gpu_macro_block) {
                        self.set_gpu_macro_block(gpu_macro_block);
                   })
                   .def("get_cpu_macro_block", [](const train_hyperparams_t& self) {
                        return self.get_cpu_macro_block();
                   })
                   .def("get_gpu_macro_block", [](const train_hyperparams_t& self) {
                        return self.get_gpu_macro_block();
                   });
}

#endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240000

ONEDAL_PY_DECLARE_INSTANTIATOR(init_model);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_train_result);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_partial_train_result);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_infer_result);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_train_ops);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_partial_train_ops);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_finalize_train_ops);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_infer_ops);
#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240000
ONEDAL_PY_DECLARE_INSTANTIATOR(init_train_hyperparameters);
#endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240000

} // namespace linear_model

ONEDAL_PY_INIT_MODULE(linear_model) {
    using namespace dal::detail;
    using namespace linear_model;
    using namespace dal::linear_regression;

    using task_list = types<task::regression>;
    auto sub = m.def_submodule("linear_model");

#ifdef ONEDAL_DATA_PARALLEL_SPMD
    ONEDAL_PY_INSTANTIATE(init_train_ops, sub, policy_spmd, task_list);
    ONEDAL_PY_INSTANTIATE(init_infer_ops, sub, policy_spmd, task_list);
#else // ONEDAL_DATA_PARALLEL_SPMD
    ONEDAL_PY_INSTANTIATE(init_train_ops, sub, policy_list, task_list);
    ONEDAL_PY_INSTANTIATE(init_infer_ops, sub, policy_list, task_list);
    ONEDAL_PY_INSTANTIATE(init_partial_train_ops, sub, policy_list, task_list);
    ONEDAL_PY_INSTANTIATE(init_finalize_train_ops, sub, policy_list, task_list);
    ONEDAL_PY_INSTANTIATE(init_model, sub, task_list);
    ONEDAL_PY_INSTANTIATE(init_train_result, sub, task_list);
    ONEDAL_PY_INSTANTIATE(init_partial_train_result, sub, task_list);
    ONEDAL_PY_INSTANTIATE(init_infer_result, sub, task_list);
#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240000
    ONEDAL_PY_INSTANTIATE(init_train_hyperparameters, sub, task_list);
#endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240000
#endif // ONEDAL_DATA_PARALLEL_SPMD

}

ONEDAL_PY_TYPE2STR(dal::linear_regression::task::regression, "regression");

} // namespace oneapi::dal::python
